acme: add LDAPDatabase backend

frasertweedale opened the pull request:
Add the LDAP schema and LDAPDatabase backend for the ACME service.

For now this is a WIP and there are still some TODOs, but it is functional so I am creating the PR to allow for review and
early feedback.

edewata commented on base/server/share/conf/schema.ldif:
```diff
@@ -687,3 +687,87 @@ attributeTypes: ( authorityKeyHost-oid NAME 'authorityKeyHost' DESC 'Authority K
+objectClasses: ( acmeChallengeHttp01-oid NAME 'acmeChallengeHttp01'
+  SUP acmeChallenge
+  STRUCTURAL
+  MUST acmeToken )
```
These object classes are identical. Would it be better to just use `acmeChallenge` and store the type in an attribute. That way if we support another challenge type we don't have to change the schema.

frasertweedale replied:
The motivation is that future challenge types could involve other data that need to be stored, so we would want a different objectclass for it, to add the new fields.  I'll make sure to explain this in the design doc.

edewata replied:
OK.

frasertweedale commented on base/acme/src/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,607 @@
+        cs.load();
+        LDAPConfig dbCfg = cs.getSubStore("internaldb", LDAPConfig.class);
+        connFactory = new LdapBoundConnFactory("acme");
+        connFactory.init(cs, dbCfg, CMSEngine.getPasswordStore("acme", cs));
```
Sure, that is a fine suggestion.  I'll make a preliminary PR for that.

frasertweedale replied:
@edewata see https://github.com/dogtagpki/pki/pull/341

frasertweedale commented on base/acme/src/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,607 @@
+        LDAPAttribute attrKey = entry.getAttribute(ATTR_ACCOUNT_KEY);
+        account.setJWK(JWK.fromJSON(attrKey.getStringValues().nextElement()));
+
+        // TODO orders ?
```
Good pickup, thanks.

frasertweedale commented on base/acme/src/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,607 @@
+
+        // In the ACME data model, an authorization may be associated with
+        // zero, one or multiple orders.  But here we ignore this and just
+        // return the first matching entry.
```
Not quite sure I follow - do you mean this method only return a pending authz?

If so I see two options:

1. Change the function to:
```java
public Collection<ACMEOrder> getOrdersByAuthorization(String authzID)
```
and return *all* orders, allowing the caller to filter it down to pending order(s).

or,

2. Change it to `getPendingOrder(s)ByAuthorization` and fliter it down to pending authorization(s) only, within this method.

What do you think?

edewata replied:
Yeah, option #2 might be better since `ACMEChallengeService` can use the results immediately without any additional filtering. Regardless, `ACMEChallengeService` will have to process multiple orders instead of just one.

Do you want me to make this change or do you prefer doing it? It will affect the other database plugins.

frasertweedale replied:
Hm, I'll have a go at it.  If I get stuck on the PostgreSQL impl I'll kick it over to you then :)

edewata commented:
@frasertweedale Please take a look at PR #337, I think it will affect this PR.

frasertweedale commented:
Now waiting on merge of #341, #342 and #343.  After that I will rebase and remove the WIP label.

frasertweedale commented:
Removed WIP label and ready for review.

edewata commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,650 @@
+                }
+            }
+            addChallenge(authorization.getAccountID(), challenge);
+        }
```
In `PostgreSQLDatabase` we delete all old challenges first, then add the new ones, but the lists of old and new challenges might be different. The above code, however, uses the same list to remove and add the challenges. In case we're removing a challenge from the authorization object, the challenge might not actually be removed from LDAP since it's not in the list. I'm not sure this is actually a scenario that we will encounter in ACME, but generally that might be considered a bug.

frasertweedale replied:
Yeah, I'll be conservative and do it the way you suggest.

edewata commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,650 @@
+        try {
+            conn.add(entry);
+        } catch (LDAPException e) {
+            throw new Exception("LDAPDatabase.ldapAdd: failed to add entry", e);
```
Could we chain the exception message, e.g.:
```
throw new Exception("LDAPDatabase.ldapAdd: failed to add entry: " + e, e);
```
Same thing with the rest of the code below.

edewata commented on base/server/share/conf/schema.ldif:
```diff
@@ -687,3 +687,92 @@ attributeTypes: ( authorityKeyHost-oid NAME 'authorityKeyHost' DESC 'Authority K
 add: objectClasses
 objectClasses: ( authority-oid NAME 'authority' DESC 'Certificate Authority' SUP top STRUCTURAL MUST ( cn $ authorityID $ authorityKeyNickname $ authorityEnabled $ authorityDN ) MAY ( authoritySerial $ authorityParentID $ authorityParentDN $ authorityKeyHost $ description ) X-ORIGIN 'user defined' )
+
+dn: cn=schema
```
Is it necessary to define the schema twice? Is there any way to avoid that?

frasertweedale replied:
There is precedent for LDAP-based profile storage and LWCAs.  I can't remember why I did it this way (possibly because the standalone file makes it easier for `ipa-server-upgrade` or human administrators to update the schema).  If we want to address it, we should file a separate ticket and do it separately.

edewata replied:
OK

frasertweedale commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,650 @@
+        String error = order.getError();
+        if (error != null) {
+            attrSet.add(new LDAPAttribute(ATTR_ERROR, error));
+        }
```
I'll add the error if it happens to be set when creating the order object. (I agree it shouldn't happen).

And I shall update `updateOrder` to set/update the error value too.

frasertweedale commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,650 @@
+        } else if (classes.contains(OBJ_CHALLENGE_HTTP01)) {
+            challenge.setType("http-01");
+        } else {
+            challenge.setType("unknown-challenge-type");  // shouldn't happen
```
OK, will throw exception.

frasertweedale commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,650 @@
+        ACMEAuthorization authz = getAuthorization(challenge.getAuthzID(), LoadChallenges.DontLoad);
+        if (authz == null) return null;
+
+        authz.setChallenges(Collections.singleton(challenge));
```
Got to construct a `Collection<ACMEChallenge>` somehow.  Seems good to me.

edewata replied:
In case we're writing a unit test that modifies the challenges in an authorization object obtained using this method the test might fail since the list of challenges is immutable. Also if the test tries to verify all challenges associated with the authorization object obtained using this method the test might fail too since the method doesn't load all challenges. It's probably not a problem now, but it could be a problem in the future.

frasertweedale replied:
Oh I didn't realise it was immutable. 

```
nobody: ...
java: let's return data that violate their own interface contacts!
```

Sigh.  I'll change it.

frasertweedale commented:
> Please see my comments below.

Thanks @edewata.  I have addressed most of your comments in the code, and the couple that I did not I have replied separately.

edewata commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,662 @@
+        String error = order.getError();
+        LDAPAttribute attrError;
+        if (error == null) {
+            attrError = new LDAPAttribute(ATTR_ERROR);  // no values
```
Will this remove the error attribute, or just blank the value?

frasertweedale replied:
I *think* it should remove the attribute.  But I should confirm (and if it does, I will put a comment).

frasertweedale replied:
Yes, confirmed.  Have elaborated in the comment.

edewata commented on base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:
```diff
@@ -0,0 +1,662 @@
+        ldapModify(dn, mods);
+
+        // Delete and re-add challenge objects.  First reload this
+        // authz so ew can delete all existing challenges.
```
Typo: ew
